File: src/diff.cpp
#include "mar/diff.hpp"
#include "mar/checksum.hpp"
#include <iostream>
#include <iomanip>
#include <sstream>

namespace mar {

namespace {
    std::string format_bytes(u64 bytes) {
        const char* units[] = {"B", "KB", "MB", "GB", "TB"};
        int unit = 0;
        double size = static_cast<double>(bytes);
        while (size >= 1024 && unit < 4) {
            size /= 1024;
            unit++;
        }
        std::stringstream ss;
        ss << std::fixed << std::setprecision(2) << size << " " << units[unit];
        return ss.str();
    }
}

DiffStatistics ArchiveDiffer::compare(const MarReader& source, const MarReader& target) {
    stats_ = DiffStatistics();
    
    // Compare file-level information
    auto source_names = source.get_names();
    auto target_names = target.get_names();
    std::unordered_set<std::string> source_names_set(source_names.begin(), source_names.end());
    std::unordered_set<std::string> target_names_set(target_names.begin(), target_names.end());

    for (const auto& name : source_names) {
        auto s_entry_opt = source.find_file(name);
        if (!s_entry_opt) continue;
        
        if (target_names_set.count(name)) {
            auto t_entry_opt = target.find_file(name);
            if (t_entry_opt) {
                const auto& s_entry = s_entry_opt->second;
                const auto& t_entry = t_entry_opt->second;
                
                bool modified = (s_entry.logical_size != t_entry.logical_size);
                if (!modified && source.has_hashes() && target.has_hashes()) {
                    auto s_hash = source.get_hash(s_entry_opt->first);
                    auto t_hash = target.get_hash(t_entry_opt->first);
                    if (s_hash && t_hash && *s_hash != *t_hash) {
                        modified = true;
                    }
                }
                
                if (modified) {
                    stats_.files_modified++;
                    stats_.bytes_modified += s_entry.logical_size;
                } else {
                    stats_.files_unchanged++;
                    stats_.bytes_unchanged += s_entry.logical_size;
                }
            }
        } else {
            stats_.files_deleted++;
            stats_.bytes_deleted += s_entry_opt->second.logical_size;
        }
    }

    for (const auto& name : target_names) {
        if (!source_names_set.count(name)) {
            auto t_entry_opt = target.find_file(name);
            if (t_entry_opt) {
                stats_.files_added++;
                stats_.bytes_added += t_entry_opt->second.logical_size;
            }
        }
    }

    return stats_;
}

void ArchiveDiffer::print_summary(std::ostream& out, const std::string& source_name, const std::string& target_name) {
    out << "\nDiff Summary: " << source_name << " -> " << target_name << "\n";
    out << std::string(70, '=') << "\n\n";
    
    out << "Files:\n";
    out << "  Added:     " << std::setw(6) << stats_.files_added << "\n";
    out << "  Deleted:   " << std::setw(6) << stats_.files_deleted << "\n";
    out << "  Modified:  " << std::setw(6) << stats_.files_modified << "\n";
    out << "  Unchanged: " << std::setw(6) << stats_.files_unchanged << "\n";
    out << "  Total:     " << std::setw(6) << (stats_.files_added + stats_.files_deleted + 
                                                stats_.files_modified + stats_.files_unchanged) << "\n";
    
    out << "\nSize Changes:\n";
    out << "  Added:     " << format_bytes(stats_.bytes_added) << "\n";
    out << "  Deleted:   " << format_bytes(stats_.bytes_deleted) << "\n";
    out << "  Modified:  " << format_bytes(stats_.bytes_modified) << "\n";
    out << "  Unchanged: " << format_bytes(stats_.bytes_unchanged) << "\n";
}

std::vector<FileDiff> ArchiveDiffer::get_file_diffs(const MarReader& source, const MarReader& target) {
    std::vector<FileDiff> diffs;
    
    auto source_names = source.get_names();
    auto target_names = target.get_names();
    std::unordered_set<std::string> source_names_set(source_names.begin(), source_names.end());
    std::unordered_set<std::string> target_names_set(target_names.begin(), target_names.end());

    // Check deleted and modified files
    for (const auto& name : source_names) {
        FileDiff diff;
        diff.path = name;
        
        auto s_entry_opt = source.find_file(name);
        if (!s_entry_opt) continue;
        
        diff.old_size = s_entry_opt->second.logical_size;
        if (source.has_hashes()) {
            auto hash = source.get_hash(s_entry_opt->first);
            if (hash) {
                diff.old_hash = hash_to_hex(hash->data(), hash->size());
            }
        }
        
        if (target_names_set.count(name)) {
            auto t_entry_opt = target.find_file(name);
            if (t_entry_opt) {
                const auto& t_entry = t_entry_opt->second;
                diff.new_size = t_entry.logical_size;
                if (target.has_hashes()) {
                    auto hash = target.get_hash(t_entry_opt->first);
                    if (hash) {
                        diff.new_hash = hash_to_hex(hash->data(), hash->size());
                    }
                }
                
                // Determine if modified
                bool modified = (diff.old_size != diff.new_size);
                if (!modified && source.has_hashes() && target.has_hashes()) {
                    modified = (diff.old_hash != diff.new_hash);
                }
                
                diff.type = modified ? FileDiff::MODIFIED : FileDiff::UNCHANGED;
                if (modified) {
                    stats_.files_modified++;
                    stats_.bytes_modified += diff.old_size;
                } else {
                    stats_.files_unchanged++;
                    stats_.bytes_unchanged += diff.old_size;
                }
            }
        } else {
            diff.type = FileDiff::DELETED;
            stats_.files_deleted++;
            stats_.bytes_deleted += diff.old_size;
        }
        
        diffs.push_back(diff);
    }

    // Check added files
    for (const auto& name : target_names) {
        if (!source_names_set.count(name)) {
            FileDiff diff;
            diff.path = name;
            diff.type = FileDiff::ADDED;
            
            auto t_entry_opt = target.find_file(name);
            if (t_entry_opt) {
                diff.new_size = t_entry_opt->second.logical_size;
                if (target.has_hashes()) {
                    auto hash = target.get_hash(t_entry_opt->first);
                    if (hash) {
                        diff.new_hash = hash_to_hex(hash->data(), hash->size());
                    }
                }
                stats_.files_added++;
                stats_.bytes_added += diff.new_size;
            }
            
            diffs.push_back(diff);
        }
    }

    return diffs;
}

void ArchiveDiffer::print_file_diffs(std::ostream& out, const std::vector<FileDiff>& diffs) {
    // Git-diff style output
    for (const auto& diff : diffs) {
        switch (diff.type) {
            case FileDiff::ADDED:
                out << "A  " << diff.path;
                if (diff.new_size > 0) out << " (" << format_bytes(diff.new_size) << ")";
                if (!diff.new_hash.empty()) out << " [" << diff.new_hash.substr(0, 8) << "]";
                out << "\n";
                break;
            case FileDiff::DELETED:
                out << "D  " << diff.path;
                if (diff.old_size > 0) out << " (" << format_bytes(diff.old_size) << ")";
                if (!diff.old_hash.empty()) out << " [" << diff.old_hash.substr(0, 8) << "]";
                out << "\n";
                break;
            case FileDiff::MODIFIED:
                out << "M  " << diff.path;
                if (diff.old_size != diff.new_size) {
                    out << " (" << format_bytes(diff.old_size) << " → " << format_bytes(diff.new_size) << ")";
                }
                if (!diff.old_hash.empty() && !diff.new_hash.empty() && diff.old_hash != diff.new_hash) {
                    out << " [" << diff.old_hash.substr(0, 8) << " → " << diff.new_hash.substr(0, 8) << "]";
                }
                out << "\n";
                break;
            case FileDiff::UNCHANGED:
                // Don't show unchanged files in diff view (like git)
                break;
        }
    }
}

} // namespace mar

File: pymar/src/diff.rs
use crate::reader::MarReader;
use std::collections::{HashMap, HashSet};

#[derive(Debug, Default, Clone)]
pub struct DiffStatistics {
    pub files_added: usize,
    pub files_deleted: usize,
    pub files_modified: usize,
    pub files_unchanged: usize,
    pub bytes_added: u64,
    pub bytes_deleted: u64,
    pub bytes_modified: u64,
    pub bytes_unchanged: u64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DiffType {
    Added,
    Deleted,
    Modified,
    Unchanged,
}

#[derive(Debug, Clone)]
pub struct FileDiff {
    pub path: String,
    pub diff_type: DiffType,
    pub old_size: u64,
    pub new_size: u64,
    pub old_hash: String,
    pub new_hash: String,
}

pub fn format_bytes(bytes: u64) -> String {
    let units = ["B", "KB", "MB", "GB", "TB"];
    let mut size = bytes as f64;
    let mut unit = 0;
    while size >= 1024.0 && unit < 4 {
        size /= 1024.0;
        unit += 1;
    }
    format!("{:.2} {}", size, units[unit])
}

pub fn compare_archives(source: &MarReader, target: &MarReader) -> (DiffStatistics, Vec<FileDiff>) {
    let mut stats = DiffStatistics::default();
    let mut diffs = Vec::new();

    let source_names: Vec<String> = source.get_names().to_vec();
    let target_names: Vec<String> = target.get_names().to_vec();
    let source_set: HashSet<String> = source_names.iter().cloned().collect();

    let mut target_map: HashMap<String, (usize, u64, String)> = HashMap::new();
    for (i, name) in target_names.iter().enumerate() {
        if let Some(entry) = target.get_file_entry(i) {
            let hash_str = target.get_hash(i).map(|h| hex::encode(&h)).unwrap_or_default();
            target_map.insert(name.clone(), (i, entry.logical_size, hash_str));
        }
    }

    // Process source files (deleted or modified or unchanged)
    for (i, name) in source_names.iter().enumerate() {
        let s_entry = match source.get_file_entry(i) {
            Some(e) => e,
            None => continue,
        };

        let old_size = s_entry.logical_size;
        let old_hash = source.get_hash(i).map(|h| hex::encode(&h)).unwrap_or_default();

        if let Some(&(_t_idx, new_size, ref new_hash)) = target_map.get(name) {
            let mut modified = old_size != new_size;
            if !modified && source.has_hashes() && target.has_hashes() && !old_hash.is_empty() && !new_hash.is_empty() {
                if old_hash != *new_hash {
                    modified = true;
                }
            }

            if modified {
                stats.files_modified += 1;
                stats.bytes_modified += old_size;
                diffs.push(FileDiff {
                    path: name.clone(),
                    diff_type: DiffType::Modified,
                    old_size,
                    new_size,
                    old_hash,
                    new_hash: new_hash.clone(),
                });
            } else {
                stats.files_unchanged += 1;
                stats.bytes_unchanged += old_size;
                diffs.push(FileDiff {
                    path: name.clone(),
                    diff_type: DiffType::Unchanged,
                    old_size,
                    new_size,
                    old_hash,
                    new_hash: new_hash.clone(),
                });
            }
        } else {
            stats.files_deleted += 1;
            stats.bytes_deleted += old_size;
            diffs.push(FileDiff {
                path: name.clone(),
                diff_type: DiffType::Deleted,
                old_size,
                new_size: 0,
                old_hash,
                new_hash: String::new(),
            });
        }
    }

    // Process added files
    for (i, name) in target_names.iter().enumerate() {
        if !source_set.contains(name) {
            if let Some(t_entry) = target.get_file_entry(i) {
                let new_size = t_entry.logical_size;
                let new_hash = target.get_hash(i).map(|h| hex::encode(&h)).unwrap_or_default();
                stats.files_added += 1;
                stats.bytes_added += new_size;
                diffs.push(FileDiff {
                    path: name.clone(),
                    diff_type: DiffType::Added,
                    old_size: 0,
                    new_size,
                    old_hash: String::new(),
                    new_hash,
                });
            }
        }
    }

    (stats, diffs)
}

pub fn print_summary(stats: &DiffStatistics, source_name: &str, target_name: &str) {
    println!("\nDiff Summary: {} -> {}", source_name, target_name);
    println!("{}", "=".repeat(70));
    println!();
    println!("Files:");
    println!("  Added:     {:>6}", stats.files_added);
    println!("  Deleted:   {:>6}", stats.files_deleted);
    println!("  Modified:  {:>6}", stats.files_modified);
    println!("  Unchanged: {:>6}", stats.files_unchanged);
    let total = stats.files_added + stats.files_deleted + stats.files_modified + stats.files_unchanged;
    println!("  Total:     {:>6}", total);
    println!();
    println!("Size Changes:");
    println!("  Added:     {}", format_bytes(stats.bytes_added));
    println!("  Deleted:   {}", format_bytes(stats.bytes_deleted));
    println!("  Modified:  {}", format_bytes(stats.bytes_modified));
    println!("  Unchanged: {}", format_bytes(stats.bytes_unchanged));
}

pub fn print_file_diffs(diffs: &[FileDiff]) {
    for diff in diffs {
        match diff.diff_type {
            DiffType::Added => {
                let mut out = format!("A  {}", diff.path);
                if diff.new_size > 0 {
                    out.push_str(&format!(" ({})", format_bytes(diff.new_size)));
                }
                if !diff.new_hash.is_empty() {
                    let short = &diff.new_hash[..diff.new_hash.len().min(8)];
                    out.push_str(&format!(" [{}]", short));
                }
                println!("{}", out);
            }
            DiffType::Deleted => {
                let mut out = format!("D  {}", diff.path);
                if diff.old_size > 0 {
                    out.push_str(&format!(" ({})", format_bytes(diff.old_size)));
                }
                if !diff.old_hash.is_empty() {
                    let short = &diff.old_hash[..diff.old_hash.len().min(8)];
                    out.push_str(&format!(" [{}]", short));
                }
                println!("{}", out);
            }
            DiffType::Modified => {
                let mut out = format!("M  {}", diff.path);
                if diff.old_size != diff.new_size {
                    out.push_str(&format!(" ({} → {})", format_bytes(diff.old_size), format_bytes(diff.new_size)));
                }
                if !diff.old_hash.is_empty() && !diff.new_hash.is_empty() && diff.old_hash != diff.new_hash {
                    let s_old = &diff.old_hash[..diff.old_hash.len().min(8)];
                    let s_new = &diff.new_hash[..diff.new_hash.len().min(8)];
                    out.push_str(&format!(" [{} → {}]", s_old, s_new));
                }
                println!("{}", out);
            }
            DiffType::Unchanged => {}
        }
    }
}

mod hex {
    pub fn encode(data: &[u8]) -> String {
        let mut s = String::with_capacity(data.len() * 2);
        for &b in data {
            s.push_str(&format!("{:02x}", b));
        }
        s
    }
}
